source/common/http/filter/gzip_filter.cc

```diff
@@ -0,0 +1,186 @@
+const uint64_t GzipFilterConfig::GZIP_HEADER_VALUE{16};
+// Default zlib memory level.
+const uint64_t GzipFilterConfig::DEFAULT_MEMORY_LEVEL{8};
+// Minimum length of an upstream response that allows compression.
+const uint64_t GzipFilterConfig::MINIMUM_CONTENT_LENGTH{30};
+
+GzipFilterConfig::GzipFilterConfig(const envoy::api::v2::filter::http::Gzip& gzip)
+    : compression_level_(compressionLevelEnum(gzip.compression_level())),
+      compression_strategy_(compressionStrategyEnum(gzip.compression_strategy())),
+      content_length_(static_cast<uint64_t>(gzip.content_length().value())),
+      memory_level_(static_cast<uint64_t>(gzip.memory_level().value())),
+      window_bits_(static_cast<uint64_t>(gzip.window_bits().value())),
+      cache_control_values_(gzip.cache_control().cbegin(), gzip.cache_control().cend()),
+      content_type_values_(gzip.content_type().cbegin(), gzip.content_type().cend()),
+      etag_(gzip.disable_on_etag().value()),
+      last_modified_(gzip.disable_on_last_modified().value()) {}
+
+ZlibCompressionLevelEnum
+GzipFilterConfig::compressionLevelEnum(const GzipV2CompressionLevelEnum& compression_level) {
+  switch (compression_level) {
+  case GzipV2CompressionLevelEnum::Gzip_CompressionLevel_Enum_BEST:
+    return ZlibCompressionLevelEnum::Best;
+  case GzipV2CompressionLevelEnum::Gzip_CompressionLevel_Enum_SPEED:
+    return ZlibCompressionLevelEnum::Speed;
+  default:
+    return ZlibCompressionLevelEnum::Standard;
+  }
+}
+
+ZlibCompressionStrategyEnum GzipFilterConfig::compressionStrategyEnum(
+    const GzipV2CompressionStrategyEnum& compression_strategy) {
+  switch (compression_strategy) {
+  case GzipV2CompressionStrategyEnum::Gzip_CompressionStrategy_RLE:
+    return ZlibCompressionStrategyEnum::Rle;
+  case GzipV2CompressionStrategyEnum::Gzip_CompressionStrategy_FILTERED:
+    return ZlibCompressionStrategyEnum::Filtered;
+  case GzipV2CompressionStrategyEnum::Gzip_CompressionStrategy_HUFFMAN:
+    return ZlibCompressionStrategyEnum::Huffman;
+  default:
+    return ZlibCompressionStrategyEnum::Standard;
+  }
+}
+
+uint64_t GzipFilterConfig::memoryLevel() const {
+  return memory_level_ > 0 ? memory_level_ : DEFAULT_MEMORY_LEVEL;
+}
+
+uint64_t GzipFilterConfig::minimumLength() const {
+  return content_length_ > 29 ? content_length_ : MINIMUM_CONTENT_LENGTH;
+}
+
+uint64_t GzipFilterConfig::windowBits() const {
+  return (window_bits_ > 0 ? window_bits_ : DEFAULT_WINDOW_BITS) | GZIP_HEADER_VALUE;
+}
+
+GzipFilter::GzipFilter(GzipFilterConfigSharedPtr config)
+    : skip_compression_{true}, compressed_data_(), compressor_(), config_(config) {}
+
+void GzipFilter::onDestroy() {}
+
+FilterHeadersStatus GzipFilter::decodeHeaders(HeaderMap& headers, bool) {
+  // TODO(gsagula): The current implementation checks for the presence of 'gzip' and if the same
+  // is followed by Qvalue. Since gzip is the only available encoding right now, order/priority of
```

Comment:
note there is also "deflate"(which isn't used much anymore and "br" (brotli) which is "up and coming", though I'm not sure how fast :)

Reply:
Sorry, comment should be more specific -> "Since gzip is the only available content-encoding in Envoy at the moment".

I haven't look at Brotli specification yet, but I saw that it's supposed to be superior to gzip which is quite impressive.